File: 0042-trapping-rain-water/0042-trapping-rain-water.py
class Solution:
    def trap(self, h: List[int]) -> int:
        stack = []
        res = 0
        for i in range(len(h)):
            while stack and h[stack[-1]] < h[i]:
                cur = stack.pop()
                if not stack:
                    break
                prev = stack[-1]
                res += (min(h[i], h[prev]) - h[cur]) * (i - prev - 1)
            stack.append(i)
        return res

File: 0042-trapping-rain-water/0042-trapping-rain-water.java
class Solution {
    public int trap(int[] h) {
        Stack<Integer> stack = new Stack<>();
        int res = 0;
        for (int i = 0; i < h.length; i++) {
            while (!stack.isEmpty() && h[stack.peek()] < h[i]) {
                int mid = stack.pop();
                if (stack.isEmpty()) {
                    continue;
                }
                int prev = stack.peek();
                res += (i - prev - 1) * (Math.min(h[i], h[prev]) - h[mid]);
            }
            stack.push(i);
        }
        return res;
    }
}